internal/collector/csr.go

```diff
@@ -0,0 +1,146 @@
+)
+
+var (
+	descCSRLabelsName          = "kube_csr_labels"
+	descCSRLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
+	descCSRLabelsDefaultLabels = []string{"csr"}
+
+	csrMetricFamilies = []metric.FamilyGenerator{
+		{
+			Name: descCSRLabelsName,
+			Type: metric.MetricTypeGauge,
+			Help: descCSRLabelsHelp,
+			GenerateFunc: wrapCSRFunc(func(j *certv1beta1.CertificateSigningRequest) *metric.Family {
+				labelKeys, labelValues := kubeLabelsToPrometheusLabels(j.Labels)
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   labelKeys,
+							LabelValues: labelValues,
+							Value:       1,
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_csr_created",
+			Type: metric.MetricTypeGauge,
+			Help: "Unix creation timestamp",
+			GenerateFunc: wrapCSRFunc(func(csr *certv1beta1.CertificateSigningRequest) *metric.Family {
+				ms := []*metric.Metric{}
+				if !csr.CreationTimestamp.IsZero() {
+					ms = append(ms, &metric.Metric{
+						LabelKeys:   []string{},
+						LabelValues: []string{},
+						Value:       float64(csr.CreationTimestamp.Unix()),
+					})
+				}
+
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_csr_condition",
+			Type: metric.MetricTypeGauge,
+			Help: "Condition of the csr",
+			GenerateFunc: wrapCSRFunc(func(csr *certv1beta1.CertificateSigningRequest) *metric.Family {
+				return &metric.Family{
+					Metrics: addCSRConditionMetrics(csr.Status),
+				}
+			}),
+		},
+	}
+)
+
+func wrapCSRFunc(f func(*certv1beta1.CertificateSigningRequest) *metric.Family) func(interface{}) *metric.Family {
+	return func(obj interface{}) *metric.Family {
+		csr := obj.(*certv1beta1.CertificateSigningRequest)
+
+		metricFamily := f(csr)
+
+		for _, m := range metricFamily.Metrics {
+			m.LabelKeys = append(descCSRLabelsDefaultLabels, m.LabelKeys...)
+			m.LabelValues = append([]string{csr.Name}, m.LabelValues...)
+		}
+
+		return metricFamily
+	}
+}
+
+func createCSRListWatch(kubeClient clientset.Interface, ns string) cache.ListWatch {
+	return cache.ListWatch{
+		ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+			return kubeClient.CertificatesV1beta1().CertificateSigningRequests().List(opts)
+		},
+		WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+			return kubeClient.CertificatesV1beta1().CertificateSigningRequests().Watch(opts)
+		},
+	}
+}
+
+// addCSRConditionMetrics generates one metric for each possible csr condition status
+func addCSRConditionMetrics(cs certv1beta1.CertificateSigningRequestStatus) []*metric.Metric {
+	approved := false
+	denied := false
+	for _, s := range cs.Conditions {
+		if s.Type == certv1beta1.CertificateApproved {
+			approved = true
+		}
+		if s.Type == certv1beta1.CertificateDenied {
+			denied = true
+		}
+	}
+
+	return []*metric.Metric{
+		&metric.Metric{
+			LabelValues: []string{"issued"},
+			Value:       boolFloat64(len(cs.Certificate) > 0),
```

Comment:
this is a state assessed based on data, not raw data, this should be done in PromQL

Reply:
@brancz So how about expose a metrics kube_csr_cert_length ?

Reply:
that works :+1: